src/lsm/compaction.zig

```diff
@@ -825,19 +832,176 @@ pub fn CompactionType(
             // TODO: If compaction is sequential, deferring manifest updates is unnecessary.
             const manifest = &compaction.context.tree.manifest;
             const level_b = compaction.context.level_b;
-            const snapshot_max = snapshot_max_for_table_input(compaction.context.op_min);
+
+            // update_snapshot_max() MUST be called before insert_table() and move_table() since
+            // it directly uses pointers to the ManifestLevel tables to perform updates.
+            // Calling insert_table() and move_table() before update_snapshot_max() will cause
+            // these pointers to become invalid.
+            compaction.update_snapshot_max();
             for (compaction.manifest_entries.slice()) |*entry| {
                 switch (entry.operation) {
                     .insert_to_level_b => manifest.insert_table(level_b, &entry.table),
-                    .update_in_level_a => manifest.update_table(level_b - 1, snapshot_max, &entry.table),
-                    .update_in_level_b => manifest.update_table(level_b, snapshot_max, &entry.table),
                     .move_to_level_b => manifest.move_table(level_b - 1, level_b, &entry.table),
                 }
             }
         }
     };
 }
 
+// Iterates over the data blocks in a level B table. References to the level B
+// tables to be iterated over are passed via the `tables` field in the context,
+// in the start() function.
+// * Assumption: This iterator works under the assumption that the references to
+// the tables are stable. For references fetched from the ManifestLevel, this
+// assumption only holds true till move_table & insert_table are called as a part
+// of apply_to_manifest. This is because move_table & insert_table are both operations
+// that insert/remove tables from the ManifestLevel segmented arrays, which could cause
+// rearrangements in the memory layout of the arrays. These rearrangements could
+// cause our table references to get invalidated.
+pub fn LevelBDataIteratorType(comptime Table: type, comptime Storage: type) type {
```

Comment:
I added it as a separate iterator since I was under the impression that Rafael is going to consume LevelDataIterator to implement scans. @batiati: could you please confirm?

Reply:
I'm ok with that, this implementation is generic enough to be used for both compaction and scans. In fact I think there is no reason to add `LevelB` to its name, we can just call it `LevelDataIterator` or even better just `LevelIterator`.

The particular case of having `tables: std.BoundedArray` as argument can also be used for scans, since we need to filter which TableInfos may contain the range we're looking for. I'm just not sure if a BoundedArray will suit scans (maybe we can abstract over a type when the time comes), but the idea seems good.

Reply:
FYI, I merged LevelBDataIterator into LevelDataIterator and changed the name to LevelDataIterator as a part of https://github.com/tigerbeetledb/tigerbeetle/pull/875/commits/f68290b8568dc9e410b6e1635a14e3573cdaa618 & https://github.com/tigerbeetledb/tigerbeetle/pull/875/commits/762412889e04a9456076c69e8121ef1620f1c97a.